[jit] lower batchmm to non-diff optimization

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #20039 dispatch and expose linear op
* #19988 split canonicalize_ops, make a decompose pass
* **#19987 lower batchmm to non-diff optimization**

Summary:
`batchmm` is actually a non differentiation optimization pass, it do the graph transformation and replacing `mm`s with `prim::BatchMMSide/Reduce`, and the registered prim ops will execute the `mm`s, this will go through autograd, not autodiff in fact, so we might can post pone the batchmm pass to right before the fusion pass, this will make the separation of differentiable optimization and non-differentiable optimization more clear, and also serve as the first step to make decompose `addmm` after the custom fusion pass

Test Plan:
Test that the graph of LSTM graph haven't changed because of this change. 

Differential Revision: [D15190356](https://our.internmc.facebook.com/intern/diff/D15190356)

wanchaol committed:
lower batchmm to non-diff optimization

wanchaol committed:
Update on "lower batchmm to non-diff optimization"

lower batchmm to non-diff optimization

gh-metadata: pytorch pytorch 19987 gh/wanchaol/1/head

wanchaol committed:
replace inlineGraph on "lower batchmm to non-diff optimization"

lower batchmm to non-diff optimization

gh-metadata: pytorch pytorch 19987 gh/wanchaol/1/head

bwasti commented on torch/csrc/jit/graph_executor.cpp:
```diff
@@ -635,16 +635,20 @@ struct GraphExecutorImpl {
     for (const auto& pass : getCustomPasses()) {
       pass(graph);
     }
+    // decomposition pass, decompose certain ops that will be used in the following
```
can you move this comment to the diff above?

wanchaol committed:
reorg code, add linear AD formula on "[jit] lower batchmm to non-diff optimization"

lower batchmm to non-diff optimization

gh-metadata: pytorch pytorch 19987 gh/wanchaol/1/head

wanchaol committed:
Update on "[jit] lower batchmm to non-diff optimization"

lower batchmm to non-diff optimization

gh-metadata: pytorch pytorch 19987 gh/wanchaol/1/head

apaszke commented:
This is not the meaning of non-differentiable optimization passes! The point was that after the differentiable optimizations the graph can be still run _with autograd enabled_, not necessarily be symbolically differentiated. Why did we change this?

wanchaol commented:
> This is not the meaning of non-differentiable optimization passes! The point was that after the differentiable optimizations the graph can be still run _with autograd enabled_, not necessarily be symbolically differentiated. Why did we change this?

@apaszke hmmm ok, but I believe the non-differentiation optimization passes can still be run with autograd enabled, the backward graph will go through `compileSpec` as well and run the fusion, and DifferentiableOp is still running autograd as it needed.
